### PR TITLE
Basic batching and changes to the enrich function.

### PR DESCRIPTION
**Pull request by ashleydavis**
I fixed a bunch of test failures that @nblumhardt  also went and fixed.

After that I did the most basic tests and implementation of batching that I could think of. This isn't a correct implementation yet and it doesn't support the flush function, but it does implement the API for batching that I proposed in #10. The simple implementation means that it doesn't yet affect any other part of the core.

I then made the changes proposed to the enrich function proposed in #11. This change reduces the expose of internal data structures to user code and it should make implementation of batching a bit simpler.

### CODE REVIEW

File: src/core/structured-log.js [hunk 1]
```diff
@@ -396,6 +396,70 @@
...
+        if (!batchOptions.timeDuration) {
+            batchOptions.timeDuration = 1000;
+        }
+
+        var batchedLogEvents = [];
+        var lastFlushTime = (new Date()).getTime();
+
+        return self.pipe(function (evt, next) {
+
+            if (batchFlushTimeout) {
```
Comment (nblumhardt) [inline src/core/structured-log.js]: Moving the declaration of `batchFlushTimeout` above this line would help readability.

**Comment by nblumhardt** [comment]
Nice compact little implementation!

The impl should hook into `end()` so that any pending batch is flushed when the logger is closed, so that events aren't lost on a controlled shutdown. `endWith()` is the mechanism for it - think there is another thread discussing this, will add some more detail there.

I think it'd be worth making the pipeline changes in the same PR, since the code as it stands isn't ready to use in real apps (without batching at the sink level, all we'd do is potentially delay log output, which may be confusing for anyone who jumps in and uses the library).

It'd also be good to make the pipeline changes at the same time, since, depending on how the batch-enabled pipeline works it may need a re-think anyway.

What do you think? Are you keen to start looking at options for how the pipeline can work with batching?

Cheers!

**Comment by ashleydavis** [comment]
I'll look into hooking it with 'end()' now.

I'm not ready to make the pipeline changes yet. I'm still looking for small changes/refactors to make that will simplify the larger change. Also the smaller changes help me understand the codebase... I don't want to rip it all apart until I full understand it ;)

**Comment by ashleydavis** [comment]
I'm having some difficulty figuring out how to move forward with this. Probably just because of my confusion over the `end` function as mentioned in #15.

**Comment by ashleydavis** [comment]
I've renamed the `end' function to `close` to make it easier to understand.

I've integrated the prototype batching feature with `close`. The batch is now flushed when the log is closed. There is a unit test that shows this working.

Can we get this pull request through now?

After this I'll continue with the more thorough overhaul required to get full batching and flush working.

**Comment by nblumhardt** [comment]
LGTM

**Pull request merged**